internal/grpcsync/pubsub_test.go

```diff
@@ -0,0 +1,225 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpcsync
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+type testSubscriber struct {
+	mu   sync.Mutex
+	msgs []string
+}
+
+func (ts *testSubscriber) OnMessage(msg interface{}) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	ts.msgs = append(ts.msgs, msg.(string))
+}
```

Comment:
Add a channel to the `testSubscriber` on notify that the callback is invoked.

```suggestion
type testSubscriber struct {
	mu   sync.Mutex
	msgs []string
	onMsgCh chan struct{}
}

func (ts *testSubscriber) OnMessage(msg interface{}) {
	ts.mu.Lock()
	defer ts.mu.Unlock()
	ts.msgs = append(ts.msgs, msg.(string))
	ts.onMsgCh <- struct{}{}
}

func newTestSubscriber() *testSubscriber {
	return &testSubscriber{onMsgCh: make(chan struct{})}
```